### PR TITLE
[Merged by Bors] - feat: port Order.FixedPoints

### PR DESCRIPTION
**Pull request by LukasMias**

### CODE REVIEW

File: Mathlib/Order/FixedPoints.lean [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+#align order_hom.is_fixed_pt_lfp OrderHom.is_fixed_pt_lfp
+
+theorem lfp_le_map {a : α} (ha : lfp f ≤ a) : lfp f ≤ f a :=
+  calc
+    lfp f = f (lfp f) := f.map_lfp.symm
+    _ ≤ f a := f.mono ha
+
+#align order_hom.lfp_le_map OrderHom.lfp_le_map
+
+theorem is_least_lfp_le : IsLeast { a | f a ≤ a } (lfp f) :=
```
Comment (mcdoll) [inline Mathlib/Order/FixedPoints.lean:96]:
```suggestion
theorem isLeast_lfp_le : IsLeast { a | f a ≤ a } (lfp f) :=
```

File: Mathlib/Order/FixedPoints.lean [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+
+theorem le_gfp {a : α} (h : a ≤ f a) : a ≤ gfp f :=
+  le_supₛ h
+#align order_hom.le_gfp OrderHom.le_gfp
+
+theorem gfp_le {a : α} (h : ∀ b, b ≤ f b → b ≤ a) : gfp f ≤ a :=
+  supₛ_le h
+#align order_hom.gfp_le OrderHom.gfp_le
+
+theorem is_fixed_pt_gfp : IsFixedPt f (gfp f) :=
```
Comment (mcdoll) [inline Mathlib/Order/FixedPoints.lean:124]:
```suggestion
theorem isFixedPt_gfp : IsFixedPt f (gfp f) :=
```

File: Mathlib/Order/FixedPoints.lean [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+theorem map_le_gfp {a : α} (ha : a ≤ gfp f) : f a ≤ gfp f :=
+  f.dual.lfp_le_map ha
+#align order_hom.map_le_gfp OrderHom.map_le_gfp
+
+theorem gfp_le_map {a : α} (ha : gfp f ≤ a) : gfp f ≤ f a :=
+  f.dual.map_le_lfp ha
+#align order_hom.gfp_le_map OrderHom.gfp_le_map
+
+theorem is_greatest_gfp_le : IsGreatest { a | a ≤ f a } (gfp f) :=
```
Comment (mcdoll) [inline Mathlib/Order/FixedPoints.lean:141]:
```suggestion
theorem isGreatest_gfp_le : IsGreatest { a | a ≤ f a } (gfp f) :=
```

File: Mathlib/Order/FixedPoints.lean [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+  f.le_lfp fun _ hb => (f.mono <| le_infₛ_iff.1 ha _ hb).trans hb
+#align order_hom.map_le_lfp OrderHom.map_le_lfp
+
+@[simp]
+theorem map_lfp : f (lfp f) = lfp f :=
+  have h : f (lfp f) ≤ lfp f := f.map_le_lfp le_rfl
+  h.antisymm <| f.lfp_le <| f.mono h
+#align order_hom.map_lfp OrderHom.map_lfp
+
+theorem is_fixed_pt_lfp : IsFixedPt f (lfp f) :=
```
Comment (mcdoll) [inline Mathlib/Order/FixedPoints.lean:85]:
```suggestion
theorem isFixedPt_lfp : IsFixedPt f (lfp f) :=
```

File: Mathlib/Order/FixedPoints.lean [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+## Main definitions
+
+* `order_hom.lfp`: The least fixed point of a bundled monotone function.
+* `order_hom.gfp`: The greatest fixed point of a bundled monotone function.
+* `order_hom.prev_fixed`: The greatest fixed point of a bundled monotone function smaller than or
+  equal to a given element.
+* `order_hom.next_fixed`: The least fixed point of a bundled monotone function greater than or
+  equal to a given element.
+* `fixed_points.complete_lattice`: The Knaster-Tarski theorem: fixed points of a monotone
+  self-map of a complete lattice form themselves a complete lattice.
```
Comment (mcdoll) [inline Mathlib/Order/FixedPoints.lean:28]:
```suggestion
* `OrderHom.lfp`: The least fixed point of a bundled monotone function.
* `OrderHom.gfp`: The greatest fixed point of a bundled monotone function.
* `OrderHom.prev_fixed`: The greatest fixed point of a bundled monotone function smaller than or
  equal to a given element.
* `OrderHom.next_fixed`: The least fixed point of a bundled monotone function greater than or
  equal to a given element.
* `fixedPoints.completeLattice`: The Knaster-Tarski theorem: fixed points of a monotone
  self-map of a complete lattice form themselves a complete lattice.
```

File: Mathlib/Order/FixedPoints.lean [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+  f.dual.map_le_lfp ha
+#align order_hom.gfp_le_map OrderHom.gfp_le_map
+
+theorem isGreatest_gfp_le : IsGreatest { a | a ≤ f a } (gfp f) :=
+  f.dual.is_least_lfp_le
+#align order_hom.is_greatest_gfp_le OrderHom.is_greatest_gfp_le
+
+theorem is_greatest_gfp : IsGreatest (fixedPoints f) (gfp f) :=
+  f.dual.is_least_lfp
+#align order_hom.is_greatest_gfp OrderHom.is_greatest_gfp
```
Comment (mcdoll) [inline Mathlib/Order/FixedPoints.lean:147]:
```suggestion
theorem isGreatest_gfp_le : IsGreatest { a | a ≤ f a } (gfp f) :=
  f.dual.is_least_lfp_le
#align order_hom.is_greatest_gfp_le OrderHom.isGreatest_gfp_le

theorem isGreatest_gfp : IsGreatest (fixedPoints f) (gfp f) :=
  f.dual.is_least_lfp
#align order_hom.is_greatest_gfp OrderHom.isGreatest_gfp
```
Sorry, I forgot to say that you need to fix the aligns as well (and there will most likely be errors down the line).

File: Mathlib/Order/FixedPoints.lean [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+#align order_hom.prev_fixed_le OrderHom.prev_fixed_le
+
+theorem le_next_fixed {x : α} (hx : x ≤ f x) : x ≤ f.nextFixed x hx :=
+  f.dual.prev_fixed_le hx
+#align order_hom.le_next_fixed OrderHom.le_next_fixed
+
+theorem next_fixed_le {x : α} (hx : x ≤ f x) {y : fixedPoints f} (h : x ≤ y) :
+    f.nextFixed x hx ≤ y :=
+  Subtype.coe_le_coe.1 <| lfp_le _ <| sup_le h y.2.le
+#align order_hom.next_fixed_le OrderHom.next_fixed_le
```
Comment (mcdoll) [inline Mathlib/Order/FixedPoints.lean:227]:
```suggestion
theorem prevFixed_le {x : α} (hx : f x ≤ x) : ↑(f.prevFixed x hx) ≤ x :=
  f.gfp_const_inf_le x
#align order_hom.prev_fixed_le OrderHom.prevFixed_le

theorem le_nextFixed {x : α} (hx : x ≤ f x) : x ≤ f.nextFixed x hx :=
  f.dual.prev_fixed_le hx
#align order_hom.le_next_fixed OrderHom.le_nextFixed

theorem nextFixed_le {x : α} (hx : x ≤ f x) {y : fixedPoints f} (h : x ≤ y) :
    f.nextFixed x hx ≤ y :=
  Subtype.coe_le_coe.1 <| lfp_le _ <| sup_le h y.2.le
#align order_hom.next_fixed_le OrderHom.nextFixed_le
```
and the corresponding changes for the next 7 theorems (I know it is very tedious).

File: Mathlib/Order/FixedPoints.lean [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+theorem map_infₛ_subset_fixed_points_le (A : Set α) (hA : A ⊆ fixedPoints f) :
+    f (infₛ A) ≤ infₛ A :=
+  le_infₛ fun _ hx => hA hx ▸ (f.mono <| infₛ_le hx)
+#align order_hom.map_Inf_subset_fixed_points_le OrderHom.map_infₛ_subset_fixed_points_le
+
+end PrevNext
+
+end OrderHom
+
+namespace FixedPoints
```
Comment (mcdoll) [inline Mathlib/Order/FixedPoints.lean:271]:
```suggestion
namespace fixedPoints
```

File: Mathlib/Order/FixedPoints.lean [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+      f.prevFixed (infₛ (Subtype.val '' s))
+        (f.map_infₛ_subset_fixed_points_le (Subtype.val '' s) fun _ ⟨x, hx⟩ => hx.2 ▸ x.2)
+    le_infₛ := fun _ _ hx => f.le_prev_fixed _ <| le_infₛ <| Set.ball_image_iff.2 hx
+    infₛ_le := fun _ _ hx =>
+      Subtype.coe_le_coe.1 <| le_trans (f.prev_fixed_le _) (infₛ_le <| Set.mem_image_of_mem _ hx) }
+
+/- porting note: mathlib3port version contained the instances as a list,
+   giving various "expected structure" errors -/
+/-- **Knaster-Tarski Theorem**: The fixed points of `f` form a complete lattice. -/
+instance : CompleteLattice (fixedPoints f) where
```
Comment (mcdoll) [inline Mathlib/Order/FixedPoints.lean:314]:
```suggestion
instance completeLattice: CompleteLattice (fixedPoints f) where
```
it is mentioned in the module docstring, so let's give it a name here as well.

File: Mathlib/Order/FixedPoints.lean [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+  __ := inferInstanceAs (SemilatticeSup (fixedPoints f))
+  __ := inferInstanceAs (CompleteSemilatticeInf (fixedPoints f))
+  __ := inferInstanceAs (CompleteSemilatticeSup (fixedPoints f))
+  top := ⟨gfp f, f.is_fixed_pt_gfp⟩
+  bot := ⟨lfp f, f.is_fixed_pt_lfp⟩
+  le_top := fun x => f.le_gfp x.2.ge
+  bot_le := fun x => f.lfp_le x.2.le
+
+
+end FixedPoints
```
Comment (mcdoll) [inline Mathlib/Order/FixedPoints.lean:325]:
```suggestion
```
One cool feature of Lean 4: you don't have to close namespaces in the end

**Comment by dupuisf** [comment]
Thanks!

bors r+

**Pull request closed without merging**